### PR TITLE
Add tests for prepared statements with subqueries

### PR DESCRIPTION
**Pull request by kokosing**
Add tests for prepared statements with subqueries

### CODE REVIEW

**Comment by kokosing** [comment]
I cherry-picked one commit from https://github.com/prestodb/presto/pull/6740

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 1]
```diff
@@ -452,30 +451,19 @@ private PlanBuilder appendApplyNode(
         return Optional.empty();
     }
 
-    private PlanBuilder createPlanBuilder(Expression expression)
+    private PlanBuilder createPlanBuilder(Node node)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: Maybe move `createPlanBuilder` method to `QueryPlanner` as static method? There is very similar method there: `planFrom`, which also does this magic:
```
        // Make field->symbol mapping from underlying relation plan available for translations
        // This makes it possible to rewrite FieldOrExpressions that reference fields from the FROM clause directly
        translations.setFieldMappings(relationPlan.getOutputSymbols());
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 2]
```diff
@@ -264,18 +264,21 @@ private PlanBuilder appendExistSubqueryApplyNode(PlanBuilder subPlan, ExistsPred
...
 
         Symbol exists = symbolAllocator.newSymbol("exists", BOOLEAN);
         subPlan.getTranslations().put(existsPredicate, exists);
+        ExistsPredicate rewrittenExistsPredicate = new ExistsPredicate(
+                subqueryPlanRoot.getOutputSymbols().get(0).toSymbolReference());
         return appendApplyNode(
                 subPlan,
                 existsPredicate.getSubquery(),
                 subqueryPlan,
-                Assignments.of(exists, existsPredicate),
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: Is that change required? We don't really care about any particular symbol

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 3]
```diff
@@ -6417,7 +6417,9 @@ public void testCorrelatedScalarSubqueriesWithScalarAggregation()
         // order by
         assertQuery(
                 "SELECT orderkey FROM orders o ORDER BY " +
-                        "(SELECT avg(i.orderkey) FROM orders i WHERE o.orderkey < i.orderkey AND i.orderkey % 10000 = 0)");
+                        "(SELECT avg(i.orderkey) FROM orders i WHERE o.orderkey < i.orderkey AND i.orderkey % 10000 = 0)" +
+                        ", orderkey LIMIT 1",
```
Comment (sopel39) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: make it single line? Line break looks odd here

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 4]
```diff
@@ -7855,6 +7856,28 @@ public void testQuantifiedComparison()
     }
 
     @Test
+    public void testPreparedStatementWithSubqueries()
+    {
+        List<QueryTemplate.Parameter> leftValues = parameter("left").of(
+                "",
```
Comment (sopel39) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: add also test for scalar subquery:
`SELECT 1 = (SELECT 1 WHERE 2 = ?)`

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/ExistsPredicate.java [hunk 5]
```diff
@@ -21,26 +21,26 @@
 public class ExistsPredicate
         extends Expression
 {
-    private final Query subquery;
```
Comment (kokosing) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/ExistsPredicate.java]: I changed this to `Node` as I assign `SymbolReference` to it. That way it is used in the same way as IN predicate where we assign `SymbolReference` to `valueList` in `ApplyNode#subqueryAssignments`

Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/ExistsPredicate.java]: Sorry, what I meant is change it to Expression and use a SubqueryExpression for the Query case

**Comment by kokosing** [comment]
Sounds ok. Will do it shortly ;)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: I can do that with subsequent PR, this PR get huge as for such simple fix I don't want to make even larger

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 3 above)
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: without line break this query also looks ugly. How now?

**Comment by kokosing** [comment]
Rebased and applied comments.

**Comment by martint** [comment]
Merged, thanks!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 6]
```diff
@@ -264,18 +264,21 @@ private PlanBuilder appendExistSubqueryApplyNode(PlanBuilder subPlan, ExistsPred
...
-        if (isAggregationWithEmptyGroupBy(subqueryPlan.getRoot())) {
+        PlanNode subqueryPlanRoot = subqueryPlan.getRoot();
+        if (isAggregationWithEmptyGroupBy(subqueryPlanRoot)) {
             subPlan.getTranslations().put(existsPredicate, BooleanLiteral.TRUE_LITERAL);
             return subPlan;
         }
 
         Symbol exists = symbolAllocator.newSymbol("exists", BOOLEAN);
         subPlan.getTranslations().put(existsPredicate, exists);
+        ExistsPredicate rewrittenExistsPredicate = new ExistsPredicate(
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: Some issues that occurred to me in this PR @martint (I approved it though).

Can it happen that there is no subquery output symbol during planning?

I am not sure exists should care about symbols. It seems that exists should care about cardinality only. You can have non-zero cardinality with empty rows.

This will limit `PruneUnreferencedOutputs` functionality on the randomly selected symbol: `subqueryPlanRoot.getOutputSymbols().get(0)`. The symbols should not be removed since it's referenced by `ExistsPredicate`. However, it's not needed to rewrite exists into count aggregation.

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: > I am not sure exists should care about symbols. It seems that exists should care about cardinality only. You can have non-zero cardinality with empty rows.

Yeah, that's a good point. We should think about how to model it better.
